src/hotspot/os/posix/signals_posix.cpp

```diff
@@ -76,22 +99,68 @@ extern "C" {
   typedef void (*sa_sigaction_t)(int, siginfo_t *, void *);
 }
 
-// For diagnostics to print a message once. see run_periodic_checks
-static sigset_t check_signal_done;
-static bool check_signals = true;
+// At various places we store handler information for each installed handler.
+//  SavedSignalHandlers is a helper class for those cases, keeping an array of sigaction
+//  structures.
+class SavedSignalHandlers {
+  // Note: NSIG can be largish, depending on platform, and this array is expected
+  // to be sparsely populated. To save space the contained structures are
+  // C-heap allocated. Since they only get added outside of signal handling
+  // this is no problem.
+  struct sigaction* _sa[NSIG];
+
+  bool check_signal_number(int sig) const {
+    assert(sig > 0 && sig < NSIG, "invalid signal number %d", sig);
+    return sig > 0 && sig < NSIG;
+  }
```

Comment:
We have

```c
 assert(sig > 0 && sig < NSIG, "invalid signal number %d", sig);
 return sig > 0 && sig < NSIG;
```

here in `SavedSignalHandlers::check_signal_number()` and

```c
#if defined(__APPLE__)
  return sig >= 1 && sig < NSIG;
#else
```

in `is_valid_signal()`.

Can we combine those, something like:

```c
// Returns true if signal number is valid.
static bool is_valid_signal(int sig) {
  assert(sig > 0 && sig < NSIG, "invalid signal number %d", sig);
  if (sig > 0 && sig < NSIG) {
    // Use sigaddset to check for signal validity.
    sigset_t set;
    sigemptyset(&set);
    if (sigaddset(&set, sig) == -1 && errno == EINVAL) {
      return false;
    } else {
      return true;
    }
  } else {
    return false;
  }
}
```

so then we can drop `SavedSignalHandlers::check_signal_number()` altogether?

Reply:
I would like to keep this. SavedSignalHandlers::check_signal_number() is just a simple bounds check to prevent memory overwriters, with an added assert for goot measure in debug builds. Its proximity to the array it guards makes this clear. SavedSignalHandlers gets only fed with know values in this file, so an assert is sufficient.

is_valid_signal() OTOH does a runtime check for signal validity. It is more expensive but only used in analysis situations. It pulls its signal number from a sigaction structure, which may be corrupt and contain whatever, so the runtime check makes more sense to me.

Reply:
In fact, SavedSignalHandler::_sa[NSIG] could be reduced in size to something more sensible, since we only ever feed "normal" signals into it. Same reason as NUM_IMPORTANT_SIGS. But we only talk a few kbyte here, so I did not bother.

Reply:
I would have preferred to have a single function to check whether the signal is valid, otherwise we have 2 separate specialized API that almost do the same thing.  Any future reader of the code will probably wonder again.

For code readability I'd prefer to have just one API, or at least refactor the more complex one in terms of the simpler one, and give them meaningful names to differentiate them, but OK.